File: 5_java/ParentesisBalanceados.java
import java.util.*;
public class ParentesisBalanceados {
  public static boolean parentesisBalanceados(String cadena) {	
    String apertura = "([{";
    String cierre = ")]}";
    Map<Character, Character> comparador = new HashMap<Character, Character>();
    comparador.put(')', '(');
    comparador.put(']', '[');
    comparador.put('}', '{');
    List<Character> pila = new ArrayList<Character>();
    for (int i = 0; i < cadena.length(); i++) {
      char caracter = cadena.charAt(i);
      if (apertura.indexOf(caracter) != -1) {
	pila.add(caracter);
      } else if (cierre.indexOf(caracter) != -1) {
	if (pila.size() == 0) {
	  return false;
	}
	if (pila.get(pila.size() -1) == comparador.get(caracter)) {
	  pila.remove(pila.size() -1);
	} else {
	  return false;
	}
      }
    }
    return pila.size() == 0;
  }

  public static void main(String[] args) {
    System.out.println(parentesisBalanceados("{[()]}"));
    System.out.println(parentesisBalanceados("{[(]}"));
  }
}

File: 7_typescript/parentesisBalanceados.ts
function parentesisBalanceados(cadena: string) {
  const apertura = "([{";
  const cierre = ")]}";
  const comparador: {[key: string]: string} = {")": "(", "]": "[", "}": "{"};
  const pila: string[] = [];
  for (const caracter of cadena) {
    if (apertura.includes(caracter)) {
      pila.push(caracter);
    } else if (cierre.includes(caracter)) {
      if (pila.length == 0) {
	return false;
      }
      if (pila[pila.length -1] === comparador[caracter]) {
	pila.pop();
      } else {
	return false;
      }
    }
  }
  return pila.length === 0;
}

console.log(parentesisBalanceados("{[()]}"));
console.log(parentesisBalanceados("{[(]}"));
